src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+                rewrite(rounding_shift_right(x, y), rounding_shift_right(x, y), is_x_wide_int_or_uint) ||
+                rewrite(rounding_shift_left(x, y), rounding_shift_left(x, y), is_x_wide_int_or_uint) ||
+                false) {
+                const Call *shift = Call::as_intrinsic(rewrite.result, {Call::rounding_shift_right, Call::rounding_shift_left});
+                internal_assert(shift);
+                bool is_saturated = op->value.as<Max>() || op->value.as<Min>();
+                Expr a = lossless_cast(op->type, shift->args[0]);
+                Expr b = lossless_cast(op->type.with_code(shift->args[1].type().code()), shift->args[1]);
+                if (a.defined() && b.defined()) {
+                    if (!is_saturated ||
+                        (shift->is_intrinsic(Call::rounding_shift_right) && can_prove(b >= 0)) ||
+                        (shift->is_intrinsic(Call::rounding_shift_left) && can_prove(b <= 0))) {
+                        return mutate(Call::make(op->type, shift->name, {a, b}, Call::PureIntrinsic));
+                    }
+                }
+            }
+            // clang-format on
+        }
+
+        if (value.same_as(op->value)) {
+            return op;
+        } else if (op->type != value.type()) {
+            return Cast::make(op->type, value);
+        } else {
+            return value;
+        }
+    }
+
+    Expr visit(const Call *op) override {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr mutated = IRMutator::visit(op);
+        op = mutated.as<Call>();
+        if (!op) {
+            return mutated;
+        }
+
+        auto rewrite = IRMatcher::rewriter(op, op->type);
+        if (rewrite(intrin(Call::abs, widening_sub(x, y)), cast(op->type, intrin(Call::absd, x, y))) ||
+            false) {
+            return rewrite.result;
+        }
+
+        if (no_overflow(op->type)) {
+            // clang-format off
+            if (rewrite(halving_add(x + y, 1), rounding_halving_add(x, y)) ||
+                rewrite(halving_add(x, y + 1), rounding_halving_add(x, y)) ||
+                rewrite(halving_add(x + 1, y), rounding_halving_add(x, y)) ||
+                rewrite(halving_add(x - y, 1), rounding_halving_sub(x, y)) ||
+                rewrite(halving_sub(x + 1, y), rounding_halving_sub(x, y)) ||
+                rewrite(shift_right(x + y, 1), halving_add(x, y)) ||
+                rewrite(shift_right(x - y, 1), halving_sub(x, y)) ||
+                rewrite(rounding_shift_right(x + y, 1), rounding_halving_add(x, y)) ||
+                rewrite(rounding_shift_right(x - y, 1), rounding_halving_sub(x, y)) ||
+                false) {
+                return mutate(rewrite.result);
+            }
+            // clang-format on
+        }
+
+        if (op->is_intrinsic(Call::widening_mul)) {
```

Comment:
I believe this whole block of code is handling cases where you widen by 4x and instead widening by 2x and adding an outer cast. If that's right, please add a comment.